docs/pages/pmd/devdocs/major_contributions/adding_new_cpd_language.md

```diff
@@ -8,16 +8,49 @@ permalink: pmd_devdocs_major_adding_new_cpd_language.html
 author: Romain PELISSE <belaran@gmail.com>
 ---
 
-If you wish CPD to parse a unsupported language, you can easily develop a new parser for CPD. All you need to is implements the following interface:
+First of all, thanks for the contribution!   
+  
+Happily for you, to add CPD support for a new language is now easier than ever!   
+> **Pro Tip**: If you wish to add a new language, there are more than 50 languages you could easily add with just an [Antlr grammar](https://github.com/antlr/grammars-v4).  
```

Comment:
It certainly is cool that so many new CPD languages can be implemented for nearly free. But I'm kind of worried about the number of dummy modules we'll end up with. CPD-only modules are generally extremely short implementation-wise: a couple of classes + a grammar. I think we don't need to have them each own a submodule, that adds noise to the root repo. Maybe we should build a single artifact with all CPD implementations for 7.0?

Reply:
I also think having so many modules will not be cool, but I'm not sure if we should attack that problem now, we will probably end up only with hyped languages such as Swift, Go, Kotlin, Ruby, etc.. I doubt people would want PMD to run over icalendar or hypertalk. But certainly you do have a point.

Reply:
@oowekyala certainly possible, but I wonder if that's the direction we want to push forward... having the grammars together in a cpd package means the generated parsers and AST end-up there too, even though we only care about the lexer.

Moreover, the downside of putting all in the same module, is that promoting a CPD-language to a full PMD-language will be more troublesome / produce a binary incompatibility (I upgrade from `pmd-cpd` 7.0.0 to 7.1.0 I may find that my language is no longer there, but on a separate pmd-*lang* 7.1.0 package); or we end up depending on `pmd-cpd` to contain the parser / lexer / nodes.

It would also be inconsistent, why would Go CPD support be in `pmd-cpd` but not for Apex / Groovy / Scala (I mention these as we are **not** using a grammar to support them, but artifacts produced by the language maintainers). If putting them al together, why in order to get cpd support for Go do I need to download Jorje, the Groovy runtime and Scala compiler as transitive dependencies?

If we do things right, it would not only be nearly free to add a CPD language, but to also have full PMD support… so I think we should probably push in that direction…

Reply:
Those are all very good points, thank you both for convincing me. I'm not annoyed enough by the profusion of modules to want to fight the dependency nightmare you describe @jsotuyod, so, let's roll with the current packaging. 

But would there at least be a way to group those cpd-only modules into a subdirectory of the root repo, without changing their maven coordinates? I mean my main point is that it's hard to browse the 30-something list of modules in the IDE when about 10 of them are CPD-only. My experience is that nearly all the time they're just noise. It's not a major concern, just a nice-to-have so to speak

Reply:
I get your point @oowekyala, maybe some IntelliJ [group module configuration](https://www.jetbrains.com/help/idea/creating-and-managing-modules.html#grouping-modules) will serve the purpose, I've never used it myself but it may be what you are looking for

Reply:
@oowekyala yes, we can use relative paths when defining modules / parents. Not the default in Maven, but doable. I'm unsure on IntelliJ (maybe @matifraga 's comment works), but Eclipse cares nothing about the project's location in the filesystem and will still show all modules at the same level unless manually grouped differently.

Reply:
@matifraga Thanks for the link, I managed to get a very nice setup with module groups. I can probably add a line of documentation about that in https://github.com/pmd/pmd/wiki/Setup-IDE#intellij-idea  

So, I guess every one can suit themselves with the module layout they like based on their IDE? Especially if eclipse doesn't care anyway, it's probably not worth changing the repo structure.

Reply:
I'm glad to hear 😄 I think sharing the IDE configuration is really useful to newcomers

Reply:
I think so too, but e.g. classpath settings can often pollute diffs... I think at least, things like run configurations for the designer can be shared, maybe also some code style settings that respect our checkstyle conventions (they really are a pain to set up). Idk how easily eclipse allows to share that, but with IntelliJ it's pretty trivial.